### PR TITLE
[REEF-1795] Implement REEF-on-Spark example

### PR DESCRIPTION
**Pull request by motus**
JIRA: [REEF-1795](https://issues.apache.org/jira/browse/REEF-1795)

### CODE REVIEW

**Comment by markusweimer** [comment]
@motus, can you squash and rebase, please? This branch contains a merge from `master`.

**Comment by motus** [comment]
@markusweimer Done!

File: lang/scala/reef-examples-scala/build.sbt [hunk 1]
```diff
@@ -0,0 +1,42 @@
+/*
```
Comment (markusweimer) [inline lang/scala/reef-examples-scala/build.sbt]: Why do we need SBT? Can't we build the whole thing with maven alone?

File: lang/scala/reef-examples-scala/build.sbt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+name         := "REEF-on-Spark"
+version      := "1.02"
```
Comment (markusweimer) [inline lang/scala/reef-examples-scala/build.sbt]: Version `1.02`?

File: lang/scala/reef-examples-scala/build.sbt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+organization := "org.apache.reef"
+
+scalaVersion := "2.11.8"
+
+publishMavenStyle := true
+
+val sparkVersion = "2.1.0"
+val reefVersion  = "0.16.0-SNAPSHOT"
```
Comment (markusweimer) [inline lang/scala/reef-examples-scala/build.sbt]: We should pick this up automatically OR add this to the version maintenance scripts we use for releases.

File: lang/scala/reef-examples-scala/pom.xml [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    <parent>
+        <groupId>org.apache.reef</groupId>
+        <artifactId>reef-project</artifactId>
+        <version>0.16.0-SNAPSHOT</version>
+        <relativePath>../../..</relativePath>
+    </parent>
+
+    <properties>
+        <rootPath>${basedir}/../../..</rootPath>
+        <findbugs.skip>true</findbugs.skip>
```
Comment (markusweimer) [inline lang/scala/reef-examples-scala/pom.xml]: Please document why we can skip findbugs in this project.

File: lang/scala/reef-examples-scala/pom.xml [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>reef-io</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <!-- End of REEF -->
+
+        <!-- Spark -->
+        <dependency>
```
Comment (markusweimer) [inline lang/scala/reef-examples-scala/pom.xml]: Move all of these new dependencies also to the `dependencyManagement` elemt of the master POM.

File: lang/scala/reef-examples-scala/pom.xml [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                    <configuration>
+                        <configLocation>lang/java/reef-common/src/main/resources/checkstyle-strict.xml</configLocation>
+                    </configuration>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+
+        <plugins>
+
+            <plugin>
```
Comment (markusweimer) [inline lang/scala/reef-examples-scala/pom.xml]: Same. Please add this to the `pluginManagement` part of the master pom.

File: lang/scala/reef-examples-scala/pom.xml [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        </dependency>
+
+    </dependencies>
+
+
+    <build>
+ 
+        <sourceDirectory>src/main/scala</sourceDirectory>
+ 
+        <pluginManagement>
```
Comment (markusweimer) [inline lang/scala/reef-examples-scala/pom.xml]: Why do we need a `pluginManagement` in a leaf POM?

File: lang/scala/reef-examples-scala/src/main/scala/org/apache/reef/examples/hellospark/ReefOnSpark.scala [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    .build
+
+  def main(args: Array[String]) {
+
+    LOG.setLevel(Level.FINEST)
+
+    val conf = new SparkConf().setAppName("ReefOnSpark:host")
+    val sc = new SparkContext(conf)
+
+    for (client <- managed(DriverLauncher.getLauncher(runtimeConfig))) {
```
Comment (markusweimer) [inline lang/scala/reef-examples-scala/src/main/scala/org/apache/reef/examples/hellospark/ReefOnSpark.scala]: Why do we need a `for` loop here?

File: lang/scala/reef-examples-scala/pom.xml (hunk 5 above)
Comment (motus) [inline lang/scala/reef-examples-scala/pom.xml]: @markusweimer Do you want to move Spark and Scala dependencies all the way to master POM? I would rather keep them in `reef-examples-scala` as long as no other modules use Scala or Spark.

File: lang/scala/reef-examples-scala/src/main/scala/org/apache/reef/examples/hellospark/ReefOnSpark.scala (hunk 8 above)
Comment (motus) [inline lang/scala/reef-examples-scala/src/main/scala/org/apache/reef/examples/hellospark/ReefOnSpark.scala]: `for`+`managed` is Scala-speak for `try`+`AutoCloseable` in Java

File: lang/scala/reef-examples-scala/pom.xml (hunk 5 above)
Comment (markusweimer) [inline lang/scala/reef-examples-scala/pom.xml]: Hmm, it is likely that we get more uses of Spark in the future. Hence, yes, I suggest to move all of this to the parent POM's `dependencyManagement` section.

**Comment by motus** [comment]
OK, I think this is the minimal `pom.xml` I can produce to make both maven and IntelliJ happy.. @markusweimer please take a look
Thanks!

**Pull request closed without merging**